ql/src/java/org/apache/hadoop/hive/ql/stats/ColStatsProcessor.java

```diff
@@ -141,14 +141,29 @@ private boolean constructColumnStatsFromPackedRows(Table tbl, List<ColumnStatist
 
       if (!statsObjs.isEmpty()) {
         if (!isTblLevel) {
-          List<FieldSchema> partColSchema = tbl.getPartCols();
+          List<FieldSchema> partColSchema = new ArrayList<>();
           List<String> partVals = new ArrayList<>();
-          // Iterate over partition columns to figure out partition name
-          for (int i = pos; i < pos + partColSchema.size(); i++) {
-            Object partVal = ((PrimitiveObjectInspector) fields.get(i).getFieldObjectInspector())
+          
+          if (tbl.hasNonNativePartitionSupport()) {
+            ObjectInspector inspector = fields.get(pos).getFieldObjectInspector();
+            if (inspector.getCategory() == ObjectInspector.Category.STRUCT) {
```

Comment:
Can `inspector.getCategory()` be anything else than `STRUCT`? What should happen if yes?

Reply:
it is always STRUCT, we compute stats per partition spec:
see `TransformSpec#toNamedStruct